main.go

```diff
@@ -36,46 +36,68 @@ type root struct {
 // StevEx: What are your thoughts on using "this" as the reciever. I've seen some differing opinions on the internet
 // also is the reciever a pointer to root? is that what r *root means? That r "is the type pointer to root" and when we call the method we are "passing"
 // the currently instantiated struct?  My main point of confusion is that in newTransaction() we seem to be passing an argument but most other methods seem to pass
-// ---------------------------------------------------------
-// commands should be methods others can be functions
-
-func repl() {
+func repl(root root) {
 
 	reader := bufio.NewReader(os.Stdin)
 	input, _ := reader.ReadString('\n')
-	parseCommand(input)
+	root.parseCommand(input)
 }
 
 func (r *root) read(key string) (string, error) {
-	if r.db[key] != 0 int {
-		return "", r.db[key]
-	}
 	displayValue := ""
+	if r.transactions == nil {
+		if value, ok := r.db[key]; ok {
+			return value, nil
+		}
+
+		if r.transactions == nil {
+			return displayValue, errors.New("That key was not found")
+		}
+	}
+
 	currentTransaction := r.transactions
+
 	for {
```

Comment:
I think this `for` loop might loop infinitely if a key is not present. Consider:

* In general, `for` will loop until you `return` from the function, `break` out of the loop, or `goto` a label.
* In this case specifically, line 68 will `return` if we have found a `write` operation in and open transaction.

Which is good! But, what about the following cases:

* The key does not exist in either the open transaction, or the root database. It seems like we'll loop forever. It seems like once we get to the last transaction (which should have a `nil` parent), we will evaluate the `if` statement on line 63, and since it's `false`, we'll go to line 72 to evaluate that if statement. Since the `currentTransaction.parent == nil`, we will not set the parent, at which point we will go back to line 60 and execute the whole loop over and over again, forever.
* The key does not exist in any open transaction, but does exist in the root database map. In this case, we'll loop forever, similar to the last case, but what we want to do is stop the loop and check the root db once `currentTransaction.parent == nil`.

So overall, I might consider the following:

1. Try using the following instead:
  ```
  for currentTransaction != nil {
    currentTransaction = currentTransaction.parent
    [...]
  ```

  This would cause the loop to terminate when the parent is nil.
2. This would also allow you to move the `if r.transactions == nil` on line 48 down below the loop, which will solve the problem of checking root.

Thoughts?